File: 백준/Silver/14501. 퇴사/퇴사.py
import sys 
input = sys.stdin.readline

n = int(input())
schedule = [()]
for _ in range(n):
    schedule.append((tuple(map(int,input().split()))))

dp = [0] * (n+1)

for i in range(1,n+1):
    if i + schedule[i][0] <= n + 1:
        dp[i] = schedule[i][1]
    for j in range(1,i):
        if i >= j + schedule[j][0] and i + schedule[i][0] <= n + 1:
            dp[i] = max(dp[i], dp[j] + schedule[i][1])
        
print(max(dp))

File: 백준/Silver/14501. 퇴사/퇴사.java
import java.io.*;
import java.util.*;

public class Main {

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		int n = Integer.parseInt(br.readLine());
		int advice[][] = new int[n][2];
		int dp[] = new int[n + 1];
		for (int i = 0; i < n; i++) {
			st = new StringTokenizer(br.readLine());
			advice[i][0] = Integer.parseInt(st.nextToken());
			advice[i][1] = Integer.parseInt(st.nextToken());
		}
		for (int i = 0; i < n; i++) {
			int idx = i + advice[i][0];
			if (idx <= n) {
				for (int j = idx; j <= n; j++)
					dp[j] = Math.max(dp[j], dp[i] + advice[i][1]);
			}
		}
		System.out.println(dp[n]);
	}
}